python/adt.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+"""
+Algebraic Data Types (ADT) is used to represent two kinds of things:
```

Comment:
"professional" "python" and "programmer" shouldn't ever be used together))

actually, they may dislike it of course, but, in this case I do not discriminate over `type`, but using the visitor. The same approach is used in python compiler libs itself, so I assumed that python programmers can hold it. What concerning, discriminating over type, `isinstance` is ubiquitous in python code base.